synthesis.go

```diff
@@ -0,0 +1,262 @@
+package poly
+
+import (
+	"github.com/jmoiron/sqlx"
+	//"github.com/juliangruber/go-intersect"
+	_ "github.com/mattn/go-sqlite3"
+	"regexp"
+	"strings"
+	"sync"
+)
+
+/******************************************************************************
+Apr 25, 2021
+
+DNA synthesis fixing stuff starts here.
+
+[Check commit history for previous comments]
+
+The primary goal of the synthesis fixer is to fix protein coding sequences (CDS) in preparation
+for DNA synthesis. Because CDSs are flexible in their coding, we can change codons to remove
+problematic parts of sequences. The synthesis fixer is built to be very fast while covering the
+majority of use-cases for synthesis fixing.
+
+Our synthesis fixer function (FixCds) takes in 3 parameters: a coding sequence, a codon table,
+and a list of functions that identify problematic sequences. These functions output not only
+the region where a problem occurs, but also some heuristics to help the function quickly find a
+solution to the given constraints. For example, function outputs can include a gc bias.
+
+The synthesis fixer works in a declarative rather than imperative approach. We spend a lot of time
+setting up our problem space so that we can quickly and easily move to a solution. First, we build
+an in-memory SQLite database with everything about our sequence - including its codon table,
+possible codon-to-codont transformations, and the sequence itself.
+
+We then run all the user-provided problematicSequenceFuncs in-parallel to generate DnaSuggestions
+of places that we should fix in our sequence. Then, we check if there are any DnaSuggestions that
+overlap - for example, perhaps a repeat region needs to be fixed in a location with a high GC content,
+so we bias fixing that repeat region to fixes that have a high AT content. If there are any overlapping
```

Comment:
just out of naive curiosity, why is having a high AT content good?

Reply:
Synthesizers like neutral 50:50 AT/GC content because that is easier for them to synthesize. If the region we want to fix has high GC content, then by adding AT biased fixes, we can equalize that AT/GC ratio to be nearer to 50:50.

Reply:
cool